### ISSUE TITLE
OSError: [Errno 101] Network is unreachable

### ISSUE THREAD

**Issue by ashish1405** [andreikop/python-ws-discovery#37]
Hello I'm getting error `OSError: [Errno 101] Network is unreachable` when I run below code

```python
from wsdiscovery.discovery import ThreadedWSDiscovery as WSDiscovery
wsd = WSDiscovery()
wsd.start()
services = wsd.searchServices()
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/ashish/anaconda3/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/ashish/.local/lib/python3.6/site-packages/wsdiscovery/threaded.py", line 150, in run
    self._sendPendingMessages()
  File "/home/ashish/.local/lib/python3.6/site-packages/wsdiscovery/threaded.py", line 226, in _sendPendingMessages
    self._sendMsg(msg)
  File "/home/ashish/.local/lib/python3.6/site-packages/wsdiscovery/threaded.py", line 213, in _sendMsg
    sock.sendto(data, (msg.getAddr(), msg.getPort()))
OSError: [Errno 101] Network is unreachable
```

but if I try to search directly like below, I'm able to get response

```python
import socket
import struct
import sys

message = '<?xml version="1.0" encoding="UTF-8"?>\n<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsdd="http://schemas.xmlsoap.org/ws/2005/04/discovery" xmlns:dn="http://www.onvif.org/ver10/network/wsdl"><SOAP-ENV:Header><wsa:MessageID>urn:uuid:5eb52dba-1787-49f8-ab8b-4567327b23c6</wsa:MessageID><wsa:ReplyTo SOAP-ENV:mustUnderstand="true"><wsa:Address>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:Address></wsa:ReplyTo><wsa:To SOAP-ENV:mustUnderstand="true">urn:schemas-xmlsoap-org:ws:2005:04:discovery</wsa:To><wsa:Action SOAP-ENV:mustUnderstand="true">http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</wsa:Action></SOAP-ENV:Header><SOAP-ENV:Body><wsdd:Probe><wsdd:Types>dn:NetworkVideoTransmitter</wsdd:Types></wsdd:Probe></SOAP-ENV:Body></SOAP-ENV:Envelope>'
multicast_group = ('239.255.255.250', 3702)

# Create the datagram socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

# Set a timeout so the socket does not block indefinitely when trying
# to receive data.
sock.settimeout(5)

# Set the time-to-live for messages to 1 so they do not go past the
# local network segment.
#ttl = struct.pack('b', 1)
#sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, ttl)

try:

    # Send data to the multicast group
    print(multicast_group)
    sent = sock.sendto(message.encode("UTF-8"), multicast_group)

    # Look for responses from all recipients
    while True:
        print ('waiting to receive',file=sys.stderr)
        try:
            data, server = sock.recvfrom(2048)
        except socket.timeout:
            print ('timed out, no more responses',file=sys.stderr)
            break
        else:
            print ('received "%s" from %s' % (data, server),file=sys.stderr)

finally:
    print ('closing socket',file=sys.stderr)
    sock.close()

```

output is

```sh
waiting to receive
received "b'<?xml version="1.0" encoding="UTF-8"?><s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:enc="http://www.w3.org/2003/05/soap-encoding" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsa5="http://www.w3.org/2005/08/addressing" xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery" xmlns:dn="http://www.onvif.org/ver10/network/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema" xmlns:tds="http://www.onvif.org/ver10/device/wsdl"><s:Header><wsa:MessageID>uuid:31d79a54-e297-2d18-6a57-f259d96f9b2f</wsa:MessageID><wsa:RelatesTo>urn:uuid:5eb52dba-1787-49f8-ab8b-4567327b23c6</wsa:RelatesTo><wsa:To>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:To><wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/ProbeMatches</wsa:Action></s:Header><s:Body><d:ProbeMatches><d:ProbeMatch><wsa:EndpointReference><wsa:Address>urn:uuid:ddaf00c9-33a2-63e9-e4ae-3ac651e662c5</wsa:Address></wsa:EndpointReference><d:Types>dn:NetworkVideoTransmitter tds:Device</d:Types><d:Scopes>onvif://www.onvif.org/Profile/Streaming onvif://www.onvif.org/location/country/china onvif://www.onvif.org/type/video_encoder onvif://www.onvif.org/name/IP-Camera onvif://www.onvif.org/hardware/HI3518C onvif://www.onvif.org/Profile/T onvif://www.onvif.org/Profile/G onvif://www.onvif.org/Profile/C onvif://www.onvif.org/Profile/A</d:Scopes><d:XAddrs>http://172.20.10.4:10001/onvif/device_service</d:XAddrs><d:MetadataVersion>1</d:MetadataVersion></d:ProbeMatch></d:ProbeMatches></s:Body></s:Envelope>'" from ('192.168.1.106', 3702)
waiting to receive
timed out, no more responses
closing socket
```

**Comment by andreikop** [comment]
Hi @ashish1405 ! 
Thank you for the report.
Could you please debug `_sendMsg` method in `NetworkingThread` at `threaded.py:212`

Replace `_sendMsg` with the next code:
```
    def _sendMsg(self, msg):
        data = createSOAPMessage(msg.getEnv()).encode("UTF-8")

        if msg.msgType() == UDPMessage.UNICAST:
            self._uniOutSocket.sendto(data, (msg.getAddr(), msg.getPort()))
            if self._capture:
                self._capture.write("%i SEND %s:%s\n" % (self._seqnum, msg.getAddr(), msg.getPort()))
                self._capture.write(data.decode("utf-8") + "\n")
                self._seqnum += 1
        else:
            for addr, sock in self._multiOutUniInSockets.items():
                pirnt("~~ Going to send message to {}:{} from socket addr {}".format(
                    msg.getAddr(), msg.getPort(), addr))
                sock.sendto(data, (msg.getAddr(), msg.getPort()))
                if self._capture:
                    self._capture.write("%i SEND %s:%s\n" % (self._seqnum, msg.getAddr(), msg.getPort()))
                    self._capture.write(data.decode("utf-8") + "\n")
                    self._seqnum += 1
```

You'll get source and destination address for message which failed to be sent.
I hope it will be clear why `send()` fails.
Looking forward for info from you.

**Comment by ashish1405** [comment]
> def _sendMsg(self, msg):
>         data = createSOAPMessage(msg.getEnv()).encode("UTF-8")
> 
>         if msg.msgType() == UDPMessage.UNICAST:
>             self._uniOutSocket.sendto(data, (msg.getAddr(), msg.getPort()))
>             if self._capture:
>                 self._capture.write("%i SEND %s:%s\n" % (self._seqnum, msg.getAddr(), msg.getPort()))
>                 self._capture.write(data.decode("utf-8") + "\n")
>                 self._seqnum += 1
>         else:
>             for addr, sock in self._multiOutUniInSockets.items():
>                 pirnt("~~ Going to send message to {}:{} from socket addr {}".format(
>                     msg.getAddr(), msg.getPort(), addr))
>                 sock.sendto(data, (msg.getAddr(), msg.getPort()))
>                 if self._capture:
>                     self._capture.write("%i SEND %s:%s\n" % (self._seqnum, msg.getAddr(), msg.getPort()))
>                     self._capture.write(data.decode("utf-8") + "\n")
>                     self._seqnum += 1

Hello @andreikop , below is the output

```
~~ Going to send message to 239.255.255.250:3702 from socket addr 172.18.0.1
~~ Going to send message to 239.255.255.250:3702 from socket addr 172.17.0.1
~~ Going to send message to 239.255.255.250:3702 from socket addr 192.168.1.106
~~ Going to send message to 239.255.255.250:3702 from socket addr 192.168.1.150
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/ashish/.local/lib/python3.6/site-packages/wsdiscovery/threaded.py", line 150, in run
    self._sendPendingMessages()
  File "/home/ashish/.local/lib/python3.6/site-packages/wsdiscovery/threaded.py", line 245, in _sendPendingMessages
    self._sendMsg(msg)
  File "/home/ashish/.local/lib/python3.6/site-packages/wsdiscovery/threaded.py", line 214, in _sendMsg
    sock.sendto(data, (msg.getAddr(), msg.getPort()))
OSError: [Errno 101] Network is unreachable

```

**Comment by ashish1405** [comment]
Hello @andreikop, one observation, if in `_createMulticastOutSocket` I replace `socket.inet_aton(addr)` with `socket.INADDR_ANY`, it seems working. So I guess binding to particular address may be the problem.

**Comment by petri** [comment]
Which Python version is this?

**Comment by andreikop** [comment]
@ashish1405 , that is `192.168.1.150` network interface?
Can you send message from this interface using netcat or other tools?
Could you write minimal script which creates a socket and sends a message?

If we send unicast message from ANY socket - the OS chooses source address using routing tables. But I'm not sure what happens if mulitcast message is sent. Probably you don't have crash because message is sent only from one of the interfaces. You can check this with Wireshark.

**Comment by ashish1405** [comment]
> On which Python version is this?

3.6.5 and 3.6.9. Tried with both

**Comment by ashish1405** [comment]
> @ashish1405 , that is `192.168.1.150` network interface?
> Can you send message from this interface using netcat or other tools?
> Could you write minimal script which creates a socket and sends a message?
> 
> If we send unicast message from ANY socket - the OS chooses source address using routing tables. But I'm not sure what happens if mulitcast message is sent. Probably you don't have crash because message is sent only from one of the interfaces. You can check this with Wireshark.

@andreikop that is IP address of a non active interface, it is only displayed when I do `ifconfig -a`. I removed that interface and it started working. So two questions
1. Why an inactive interface is taken while sending multicast packets?
2. Even if one interface fails, shouldn't the result from other interfaces be returned?

BTW, I added/removed the interface using below if you want to reproduce.

```sh
sudo ip link add eth0 type dummy
sudo ip addr add 192.168.1.150/24 dev eth0
sudo ifconfig eth0 up
sudo ifconfig eth0 down
#tested here not working
sudo ip link del eth0 
#tested here working
```

**Comment by petri** [comment]
@ashish1405 what OS are you on? 

Regarding your questions, I don't know if it's even possible to, in this case, choose between active & inactive interfaces. It might be Python socket access implementation detail or a bug there. But I could be wrong. 

Regarding the second question, perhaps. And there might be a result that's just not shown because the error is triggered first. I have no idea.

Could you try removing the IP address from your inactive interface? Does that make any difference?

**Comment by ashish1405** [comment]
@andreikop I've already said in my previous response that **I removed that interface and it started working**. I'll try to find out answer to the two questions in my free time and post it here. If you get the same please post it here.

**Comment by andreikop** [comment]
I think good enough solution is to catch this exact exception and print warning or debug log.
@ashish1405 , maybe you could make a fix?
Just make sure only "Network in unreachable" error message is caught, but other errors still crash the application loudly.

**Comment by petri** [comment]
> @petri I've already said in my previous response that **I removed that interface and it started working**.

Yes I know. I asked about removing the IP because who knows maybe that might be enough for Python to not try the interface. No other reason.

**Comment by ashish1405** [comment]
@petri Yes I just checked that. Just removing the ip address also it works even if the interface is there.

**Comment by nikolaevigor** [comment]
I had same problem recently. My solution was to remove all interfaces, that had problematic `ip`s.
As I see `threaded.py` failing loudly if it experience at least one bad ip in `threaded.py:212` `self._multiOutUniInSockets.items()`.
Could it be better to just log bad ip with warning and continue working as @andreikop has suggested in this [comment](https://github.com/andreikop/python-ws-discovery/issues/37#issuecomment-627782589)?

**Comment by andreikop** [comment]
@nikolaevigor, your PR is welcome!

**Comment by tom24** [comment]
This looks similar to the issue I had and reported in issue #41.  If a sendto causes an error (for various reasons), then the app gives up, even though sendto's to other interfaces should still be ok.

I gave a simple fix in #41. It could be improved -- instead of 'pass' in the except case, log the reason for the exception.  Maybe logging should only be at debugging level since the environment would be such that the problem interface is necessary and do not want log clutter.

Another approach would be to check link flags (UP, MULTICAST, LOOPBACK etc) and not sendto links that do not support it.

**Comment by ashish1405** [comment]
@tom24 , a good fix, I think. You can generate a pull up request for the change.

**Comment by andreikop** [comment]
Fixed by #42
